src/gui/guiFormSpecMenu.cpp

```diff
@@ -4228,153 +4290,370 @@ bool GUIFormSpecMenu::OnEvent(const SEvent& event)
+						a->to_i = ds.first.i;
+						m_invmgr->inventoryAction(a);
+					}
+
+				} else if (identical) {
+					// Put the selected item back where it came from
+					m_selected_amount = 0;
+				} else if (s.isValid()) {
+					// Move the selected item
+					move_amount = m_selected_amount;
 				}
+
+				m_left_drag_stacks.clear();
+			}
+			break;
+		}
+		case BET_MOVE: {
+			// Mouse button is down and mouse pointer entered a new inventory field
+
+			if (!s.isValid() || s.listname == "craftpreview")
+				break;
+
+			if (!m_selected_item && event.MouseInput.Shift) {
+				// Shift-move items while dragging
+				if (m_held_mouse_button == BET_RIGHT)
+					shift_move_amount = 1;
+				else if (m_held_mouse_button == BET_MIDDLE)
+					shift_move_amount = MYMIN(s_count, 10);
+				else if (m_held_mouse_button == BET_LEFT)
+					shift_move_amount = s_count;
+
+			} else if (m_selected_item) {
+				if (m_held_mouse_button != BET_LEFT) {
+					// Move items if the destination slot is empty
+					// or contains the same item type as what is going to be moved
+					if (!m_selected_dragging && (empty || matching)) {
+						if (m_held_mouse_button == BET_RIGHT)
+							move_amount = 1;
+						else if (m_held_mouse_button == BET_MIDDLE)
+							move_amount = MYMIN(m_selected_amount, 10);
+					}
+
+				} else if (m_left_dragging && (empty || matching) &&
+						   m_left_drag_amount > m_left_drag_stacks.size()) {
+					// Add the slot to the left-drag list if it doesn't exist
+					bool found = false;
+					for (auto &ds : m_left_drag_stacks) {
+						if (s == ds.first) {
+							found = true;
+							break;
+						}
+					}
+					if (!found) {
+						m_left_drag_stacks.emplace_back(s, list_s->getItem(s.i));
+					}
+
+				} else if (m_selected_dragging && matching && !identical) {
+					// Pickup items of the same type while dragging
+					pickup_amount = s_count;
+				}
+
+			} else if (m_held_mouse_button == BET_LEFT) {
+				// Start picking up items
+				m_selected_item = new GUIInventoryList::ItemSpec(s);
+				m_selected_amount = s_count;
+				m_selected_dragging = true;
 			}
-		break;
+			break;
+		}
+		case BET_OTHER: {
+			// Some other mouse event has occured
+			// Currently only left-double-click should trigger this
+			if (!s.isValid() || event.MouseInput.Event != EMIE_LMOUSE_DOUBLE_CLICK)
+				break;
+
+			// Abort left-dragging
+			m_left_dragging = false;
+			m_client->inhibit_inventory_revert = false;
+			m_left_drag_stacks.clear();
+
+			// Both the selected item and the hovered item need to be checked
+			// because we don't know exactly when the double-click happened
+			ItemStack slct;
+			if (!m_selected_item && !empty)
+				slct = list_s->getItem(s.i);
+			else if (m_selected_item && (identical || empty))
+				slct = list_selected->getItem(m_selected_item->i);
+
+			// Pickup all of the item from the list
+			if (slct.count > 0) {
+				for (u32 i = 0; i < list_s->getSize(); i++) {
+					// Skip the selected slot
+					if (i == s.i)
```

Comment:
`src/gui/guiFormSpecMenu.cpp:4562:47: warning: comparison of integer expressions of different signedness: ‘irr::u32’ {aka ‘unsigned int’} and ‘irr::s32’ {aka ‘int’} [-Wsign-compare]`

(same as above)